### ISSUE TITLE
windows build failed with current master

### ISSUE THREAD

**Issue by wolftobias** [openssl/openssl#3778]
I just downloaded the openssl-master to try the admission feature, but the windows build fails, Linux 32/64 was fine. I tried afterwards the version openssl-1.1.0f which was fine. This leads me to the question: Is something wrong in the master build?

```
make[2]: Entering directory `/usr/local/gematik/openssl-master'
( :; LIBDEPS="${LIBDEPS:-test/libtestutil.a libcrypto.a -lws2_32 -lgdi32 -lcrypt32 }"; LDCMD="${LDCMD:-x86_64-w64-mingw32-gcc}"; LDFLAGS="${LDFLAGS:--DDSO_WIN32 -DNDEBUG -DOPENSSL_THREADS -DOPENSSL_NO_STATIC_ENGINE -DOPENSSL_PIC -DOPENSSL_IA32_SSE2 -DOPENSSL_BN_ASM_MONT -DOPENSSL_BN_ASM_MONT5 -DOPENSSL_BN_ASM_GF2m -DSHA1_ASM -DSHA256_ASM -DSHA512_ASM -DRC4_ASM -DMD5_ASM -DAES_ASM -DVPAES_ASM -DBSAES_ASM -DGHASH_ASM -DECP_NISTZ256_ASM -DPADLOCK_ASM -DPOLY1305_ASM -DOPENSSL_API_COMPAT=0x10100000L -DOPENSSLDIR="\"/usr/local/gematik/bin/openssl-master/win64/ssl\"" -DENGINESDIR="\"/usr/local/gematik/bin/openssl-master/win64/lib/engines-1_1\"" -DL_ENDIAN -DWIN32_LEAN_AND_MEAN -DUNICODE -D_UNICODE -m64 -Wall -O3 -D_MT   }"; LIBPATH=`for x in $LIBDEPS; do echo $x; done | sed -e 's/^ *-L//;t' -e d | uniq`; LIBPATH=`echo $LIBPATH | sed -e 's/ /:/g'`; echo LD_LIBRARY_PATH=$LIBPATH:$LD_LIBRARY_PATH ${LDCMD} ${LDFLAGS} -o ${APPNAME:=test/asn1_internal_test.exe} test/asn1_internal_test.o ${LIBDEPS}; LD_LIBRARY_PATH=$LIBPATH:$LD_LIBRARY_PATH ${LDCMD} ${LDFLAGS} -o ${APPNAME:=test/asn1_internal_test.exe} test/asn1_internal_test.o ${LIBDEPS} )
LD_LIBRARY_PATH=: x86_64-w64-mingw32-gcc -DDSO_WIN32 -DNDEBUG -DOPENSSL_THREADS -DOPENSSL_NO_STATIC_ENGINE -DOPENSSL_PIC -DOPENSSL_IA32_SSE2 -DOPENSSL_BN_ASM_MONT -DOPENSSL_BN_ASM_MONT5 -DOPENSSL_BN_ASM_GF2m -DSHA1_ASM -DSHA256_ASM -DSHA512_ASM -DRC4_ASM -DMD5_ASM -DAES_ASM -DVPAES_ASM -DBSAES_ASM -DGHASH_ASM -DECP_NISTZ256_ASM -DPADLOCK_ASM -DPOLY1305_ASM -DOPENSSL_API_COMPAT=0x10100000L -DOPENSSLDIR="/usr/local/gematik/bin/openssl-master/win64/ssl" -DENGINESDIR="/usr/local/gematik/bin/openssl-master/win64/lib/engines-1_1" -DL_ENDIAN -DWIN32_LEAN_AND_MEAN -DUNICODE -D_UNICODE -m64 -Wall -O3 -D_MT -o test/asn1_internal_test.exe test/asn1_internal_test.o test/libtestutil.a libcrypto.a -lws2_32 -lgdi32 -lcrypt32
libcrypto.a(mem.o):mem.c:(.text+0xb): undefined reference to `random'
collect2: error: ld returned 1 exit status
make[2]: *** [link_app.cygwin] Error 1
make[2]: Leaving directory `/usr/local/gematik/openssl-master'
make[1]: *** [test/asn1_internal_test.exe] Error 2
make[1]: Leaving directory `/usr/local/gematik/openssl-master'
make: *** [all] Error 2
remove temp source dir
[root@gematikserver gematik]#
```

**Comment by levitte** [comment]
Ah, the Windows libraries don't provide `random()`, only `rand()`.  The latter has implementations that are superbly weak in the high bits, but considering we're only looking at the lower bits in this particular case, we could as well use `rand()`.

**Comment by levitte** [comment]
Would you mind trying this out?

``` diff
diff --git a/crypto/mem.c b/crypto/mem.c
index 0584814f73..0607544718 100644
--- a/crypto/mem.c
+++ b/crypto/mem.c
@@ -116,6 +116,9 @@ static void parseit(void)
  */
 static int shouldfail(void)
 {
+#ifdef _WIN32
+# define random() rand()
+#endif
     int roll = (int)(random() % 100);
     int shoulditfail = roll < md_fail_percent;
     char buff[80];
```

**Comment by wolftobias** [comment]
Yes no problem. But what shall I do? Did you commited a change to master branch?

**Comment by mattcaswell** [comment]
Strange that AppVeyor is green??

**Comment by levitte** [comment]
@wolftobias, please apply that diff to your copy and confirm that it solved your issue.

@mattcaswell, not at all...  the code that calls `random()` is guarded, you need to `enable-crypto-mdebug` to open it up.

**Comment by wolftobias** [comment]
I made the change you asked me for directly in the source file 'mem.c' and now the build works and runs successfully!!!
When do you make the change in the master?

**Comment by levitte** [comment]
Fix is up in #3778
